lib/warden/strategies/user_basic_auth.rb

```diff
@@ -0,0 +1,30 @@
+require 'warden/basic_auth'
+
+module Warden
+  module Strategies
+    ##
+    # Allows users to authenticate using their API key via basic auth.
+    # Note that in order for a user to be able to generate one
+    # `Setting.rest_api_enabled` has to be `1`.
+    #
+    # The basic auth credentials are expected to contain the literal 'apikey'
+    # as the user name and the API key as the password.
+    class UserBasicAuth < BasicAuth
+      def self.user
+        'apikey'
```

Comment:
As `GlobalBasicAuth` takes precedence  and is tried first, then (if I'm not mistaken) `UserBasicAuth` will fail if someone is ignorant enough to use *apikey* as the global username.
You could check the global username and prevent that value from being taken.